### ISSUE TITLE
pointed_thing_to_face_pos behaves incorectly for nodes with nodebox type other then regular

### ISSUE THREAD

**Issue by Skamiz** [minetest/minetest#9147]
##### Minetest version

```
5.1.0
```

##### Summary
pointed_thing_to_face_pos() works only reliably if the pointed surface is allingned with the borders of a full node like default:stone

If the pointed suface is inside a node boundary like on the inner side of stairs and slabs the return value is often incorect.
The less perpendicular the view direction to the pointed surface the further the returned position is from the pointed node, sometimes being even several nodes behind the player.

I have since found a better way to acomplish what I was trying to do when this sprung on me, so my main issue with this, is that it isn't documented. From the function desciption and the documentation on pointed_thing I expected that it would give me the intersection point of my look direction and the pointed node as if it were a full node.

**Comment by Skamiz** [comment]
Attached file is a minimod which adds an item called axe of debug. Rightclick with it on a node to visualize the position returned by pointed_thing_to_face_pos

[face_pos_demo.zip](https://github.com/minetest/minetest/files/3887683/face_pos_demo.zip)

**Comment by SmallJoker** [comment]
The problem here is missing information in `pointed_thing`. It's not possible to get the exact face on a nodebox when only entire node positions are provided (`{ above = <pos>, under = <pos> }`).

Consider using Raycast if you want to get a precise intersection point.

**Comment by paramat** [comment]
The feature was written assuming full cubes, so this is a request for better documentation and possibly a feature request (?), not a bug.

**Comment by Skamiz** [comment]
Since using the function such as it is curently documented can produce wastly incorrect results I would definitively consider it a bug.
I don't think that wether it is fixed through corretion of the documentation or code or maybe both changes that.

I mean, it come down to expectaions, with what the curent documentation tells you it is unexpected incorrect behavior, which is obviously bad because it leads to issues appearing, in a modders code, which they didn't even cause themselves.
If you just change the documentation, then it will (given people will actualy read it) become expected incorrect, behavior. It will still create compleatly wrong results in some situations, but people will know to avoid it and in most situations raycasts will replace it easily anyway.

And then there are various options of what you could do with the code, which would reqire various chenges to the documentation.
It could just display a warning when an absurd result id detected.
It could clamp the return position so that it is still at least touching the cuboid of the pointed node (and not floating several nodes *behind* the player). Which would still be incorrect but much less so.
It could be chaged to retun the exact location on the cuboid. With a raycast and some linear interpolation I don't think taht would be very comlicated.
Or it could be changed to return the position goten from the raycast. The documentation on the function doesn't actually make any claim to it being the possition on a cuboid. That was just an inference from the documentation on pointed_thing.
Specificaly it says
`returns the exact position on the surface of a pointed node`
which by itself actualy implies that the raycast result is the "correct" one.

Choose and pick whatever resolution you think is most apropriate.
My point beeing, regardles of which aproach is choosen in the end, the current bahavior is an issue that will cause errors in peoples code even though they used the provided tools as instructed. Thus I consider this a bug. The main issue being with the documentation was a personal coment, since I relly on the lua api heavily.

**Comment by paramat** [comment]
Yes good point, it comes down to whether imperfect documentation is considerd a 'bug', i do not because a 'bug' to me means broken code. Strictly, text cannot have a 'bug'. But i can understand why others may consider that a 'bug'. This situation arises often so i will add a new label.

**Comment by paramat** [comment]
I prefer leaving the code as it is and changing the documentation to make clear this returns a position on the cubic border of a node.
I can imagine a nodebox being used where some faces are co-planar with the cubic borders and where the function is still useful to get the pointed position on those faces.
Likewise for all other drawtypes, even if no node faces are co-planar with the cubic border.

**Comment by SmallJoker** [comment]
I think it needs better documentation. State that it's a fast function that is limited to `drawtype = "normal"` nodes.
Refer to Raycast for more precise results:
https://github.com/minetest/minetest/blob/4b6bff46e14c6215828da5ca9ad2cb79aa517a6e/doc/lua_api.txt#L1343-L1353

**Comment by Wuzzy2** [comment]
It's important to be precise.
Is it really limited to `drawtype = "normal"` or do you actually mean `selection_box = { type = "regular" }` (=default 1×1×1 cube selection box)?

**Comment by paramat** [comment]
It is not limited to "normal" drawtype, as it could be used as i described in my previous comment.
It also does not depend on a selectionbox matching the cubic border of a node.
Simply this:
> this returns a position on the cubic border of a node.
